### PR TITLE
[doc] Editorial updates to contributor_guide.md

### PR DESCRIPTION
**Pull request by Vissidarte-Herman**
Miscellaneous editorial updates.

Related issue = #

### CODE REVIEW

**Comment by Vissidarte-Herman** [comment]
Redesigned the contribution guidelines. Done finally.

File: docs/lang/articles/contribution/contributor_guide.md [hunk 1]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
-[[PDF]](https://github.com/yuanming-hu/public_files/blob/master/graphics/taichi/google_review_comments.pdf)
 
-- PR authors are required to **squash commits on your own**.
-- If you are making multiple PRs,
-  - Independent PRs should be based on **different** branches forking from `master`;
-  - PRs with dependencies should be raised only after all prerequisite PRs are merged into `master`.
+### PR naming conventions
 
-### **PR naming conventions**
+Your PR will make it into the commit history in the the master branch or even Taichi's release notes, therefore it is important to keep your PR title self-explanatory. This section describes our PR naming conventions:
```
Comment (rexwangcc) [inline docs/lang/articles/contribution/contributor_guide.md:204]:
```suggestion
Your PR will make it into the commit history in the the master branch and even Taichi's release notes, therefore it is important to keep your PR title self-explanatory. This section describes our PR naming conventions:
```
I figure it's deterministic?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 2]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
+```
 
-We do appreciate all kinds of contributions, yet we do not intend to expose the details of every PR to the end-users. Therefore the changelog helps distinguish *what the user should know* from *what the*
+- **Tag naming conventions:**
+  - Prepend at least one tag, such as `[lang]`, to your PR title.
+  - If you have multiple tags, separate adjacent tags with one space.
+  -
+  - See [misc/prtags.json](https://github.com/taichi-dev/taichi/blob/master/misc/prtags.json) for a full list of available tags.
+  - We differentiate PRs for end-users from PRs for developers by *capitalizing tag initial*.
+    - If a PR deals with a feature visible to the end-users, initialize the most relevant tag and the PR will [make it into the release notes](ttps://github.com/taichi-dev/taichi/blob/master/misc/make_changelog.py). For example, `[Metal]`, `[Vulkan]`, `[IR]`, `[Lang]`, or `[CUDA]` Ensure that your PR title has *AT MOST* one tag dealt this way.
```
Comment (rexwangcc) [inline docs/lang/articles/contribution/contributor_guide.md:228]:
```suggestion
    - If a PR deals with a feature visible to the end-users, initialize the most relevant tag and the PR will [make it into the release notes](https://github.com/taichi-dev/taichi/blob/master/misc/make_changelog.py). For example, `[Metal]`, `[Vulkan]`, `[IR]`, `[Lang]`, or `[CUDA]` Ensure that your PR title has *AT MOST* one tag dealt this way.
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 3]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
-- A ✔️ on the left-hand side of a commit hash means that CI passes, whilst a ❌ means that CI fails.
+Following are some frequently used tags:
 
-### **PR reviewing & merging**
+- `[cuda]`: Backend-specific changes.
+- `[lang]`: Frontend language features, including syntax sugars.
+- `[ir]`: Intermediate representation-specific changes.
+- `[refactor]`: Code refactoring changes.
+- `[ci]`: CI/CD workflow-specific changes.
+- `[Doc]: Documentation updates.
```
Comment (rexwangcc) [inline docs/lang/articles/contribution/contributor_guide.md:269]:
```suggestion
- `[Doc]`: Documentation updates.
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 4]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
+Folow this checklist during PR review or merging:
 
+1. Ensure that your PR title follows our [naming conventions](#pr-naming-conventions).
+2. Ensure that Taichi's master branch has a *linear history*. See the [Linear vs Non-Linear History](https://idiv-biodiversity.github.io/git-knowledge-base/linear-vs-nonlinear.html) for more information.
+3. Ensure that your PR passes all Continuous Integration (CI) tests before merging it.
+   CI is triggered each time you push a commit to an open PR. It builds and tests all commits in your PR in multiple environments. Keep an eye on the CI test results:
+   - A ✔️ on the left-hand side of a commit hash: CI has passed,
+   - A ❌ on the left-hand side of a commit hash: CI has failed.
 
+Here, we do not want to repeat some best practices summarized in the following Google blog articles. But do please spare a couple of minutes reading them if your PR is being reviewed or if you are reviewing a PR. They have our recommendation!
```
Comment (rexwangcc) [inline docs/lang/articles/contribution/contributor_guide.md:286]:
```suggestion
Here, we do not want to repeat some best practices summarized in the following Google blog articles. But please do spare a couple of minutes reading them if your PR is being reviewed or if you are reviewing a PR. They have our recommendation!
```

**Review by rexwangcc — COMMENTED** [review]
Thanks for the updates, I left a few comments & questions :octocat:

**Review by k-ye — APPROVED** [review]

File: docs/lang/articles/contribution/contributor_guide.md [hunk 5]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
+  - [Code Health: Understanding Code In Review](https://testing.googleblog.com/2018/05/code-health-understanding-code-in-review.html)
+  - [Code Health: Respectful Reviews == Useful](https://testing.googleblog.com/2019/11/code-health-respectful-reviews-useful.html)
+  - [How to have your PR merged quickly](https://testing.googleblog.com/2017/06/code-health-too-many-comments-on-your.html)
 
 ## Frequently asked questions
 
 #### What if I didn't format my code style locally?
```
Comment (strongoier) [inline docs/lang/articles/contribution/contributor_guide.md:293]: This part seems duplicating https://github.com/Vissidarte-Herman/taichi/blob/d68a64521a625eb471245c7e94fd18ee09a18ed0/docs/lang/articles/contribution/contributor_guide.md?plain=1#L115-L120.

**Review by strongoier — CHANGES_REQUESTED** [review]

File: docs/lang/articles/contribution/contributor_guide.md [hunk 6]
```diff
@@ -78,35 +74,34 @@ Except for minor updates, most PRs start from a developer taking over an issue.
...
-    2. Hold off until a core developer responds to your action plan.
-
+      1. Briefly describe how your plan to handle it (if no solution has been provided).
+      2. Hold off until a core developer responds to your action plan.
   - **Goal**: Keep your implementation neat and effective.
   - **Example**:  See [#2610](https://github.com/taichi-dev/taichi/issues/2610).
 
 ## References for documentation updates
 
+As part of the effort to increase visibility of the community and to improve developer experience, we highly recommend including documentation updates in your PR if applicable. Here are some of the documentation-specific references and tips.
```
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:86]: Yes. Should be a comma.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 5 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:293]: Yes. It is a duplicate one. I'll remove it.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 1 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:204]: all PRs will enter the commit history, but only PRs with capitalized tags will enter release notes. So, it ought to be "or" here.

File: docs/lang/articles/contribution/contributor_guide.md [hunk 7]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
 > For more style information for your C++ code, see [our C++ style](./cpp_style).
 
-## Run integration tests
-
-To run all the C++ and Python tests: Run `python tests/run_tests.py` to run all the C++ and Python tests.
+### Run integration tests
 
-**Syntax:**
+To run all the C++ and Python tests:
+```python tests/run_tests.py```
```
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:131]: All fixed.

File: docs/lang/articles/contribution/contributor_guide.md [hunk 8]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
+1. Ensure that your PR title follows our [naming conventions](#pr-naming-conventions).
+2. Ensure that Taichi's master branch has a *linear history*. See the [Linear vs Non-Linear History](https://idiv-biodiversity.github.io/git-knowledge-base/linear-vs-nonlinear.html) for more information.
+3. Ensure that your PR passes all Continuous Integration (CI) tests before merging it.
+   CI is triggered each time you push a commit to an open PR. It builds and tests all commits in your PR in multiple environments. Keep an eye on the CI test results:
+   - A ✔️ on the left-hand side of a commit hash: CI has passed,
+   - A ❌ on the left-hand side of a commit hash: CI has failed.
 
+Here, we do not want to repeat some best practices summarized in the following Google blog articles. But do please spare a couple of minutes reading them if your PR is being reviewed or if you are reviewing a PR. They have our recommendation!
+  - [Code Health: Understanding Code In Review](https://testing.googleblog.com/2018/05/code-health-understanding-code-in-review.html)
+  - [Code Health: Respectful Reviews == Useful](https://testing.googleblog.com/2019/11/code-health-respectful-reviews-useful.html)
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:288]:
```suggestion
  - [Code Health: Respectful Reviews == Useful Reviews](https://testing.googleblog.com/2019/11/code-health-respectful-reviews-useful.html)
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 9]
```diff
@@ -78,35 +74,34 @@ Except for minor updates, most PRs start from a developer taking over an issue.
...
+- For instructions on setting up a local server and previewing your updated documentation in real-time, see the [Local Development](https://github.com/taichi-dev/docs.taichi.graphics#local-development).
 
-## Add test cases for testing your local changes
+## Add test cases for your local changes
 
-We highly recommend that you write your own test cases to cover corner cases for your codes before filing a PR.
+If your PR is to implement a new feature, we recommend that you write your own test cases to cover corner cases for your codes before filing a PR.
 
 - To write a Python test case, see the [Workflow for writing a Python test](./write_test).
 - To write a C++ test case, see the [Workflow for writing a CPP test](./writing_cpp_tests).
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:98]:
```suggestion
- To write a C++ test case, see the [Workflow for writing a C++ test](./writing_cpp_tests).
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 10]
```diff
@@ -131,143 +125,175 @@ python misc/code_format.py
...
+:::
 
-- Ensure that the PR passes all **continuous integration (CI) tests**. See [Using continuous integration](./contributor_guide#using-continuous-integration).
+### PR review & merging checklist
 
-- Ensure that the title follows [PR tag rules](./contributor_guide#pr-title-format-and-tags).
+Folow this checklist during PR review or merging:
 
+1. Ensure that your PR title follows our [naming conventions](#pr-naming-conventions).
+2. Ensure that Taichi's master branch has a *linear history*. See the [Linear vs Non-Linear History](https://idiv-biodiversity.github.io/git-knowledge-base/linear-vs-nonlinear.html) for more information.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:280]:
```suggestion
2. Ensure that Taichi's master branch has a *linear history*. See [Linear vs Non-Linear History](https://idiv-biodiversity.github.io/git-knowledge-base/linear-vs-nonlinear.html) for more information.
```

**Review by yuanming-hu — COMMENTED** [review]
Some nits :-) Sorry, I did't realize this part is under construction so I filed https://github.com/taichi-dev/taichi/pull/3815 - here I moved some modifications there to your PR.

File: docs/lang/articles/contribution/contributor_guide.md [hunk 11]
```diff
@@ -69,44 +65,43 @@ Except for minor updates, most PRs start from a developer taking over an issue.
...
+If your PR is to implement a new feature, we recommend that you write your own test cases to cover corner cases for your codes before filing a PR.
 
 - To write a Python test case, see the [Workflow for writing a Python test](./write_test).
-- To write a C++ test case, see the [Workflow for writing a CPP test](./writing_cpp_tests).
+- To write a C++ test case, see the [Workflow for writing a C++ test](./writing_cpp_tests).
 
 ## Conduct formatting and integration tests locally
 
-We recommend that you complete code formatting and integration tests on your local computer before filing a PR.
+We highly recommend that you complete code formatting and integration tests on your local computer before filing a PR.
```
Comment (ailzhang) [inline docs/lang/articles/contribution/contributor_guide.md:102]: maybe more specific: complete code style checks?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 12]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
+- **If you make multiple commits to an open PR:**
 
-- PR authors are required to **squash commits on your own**.
-- If you are making multiple PRs,
-  - Independent PRs should be based on **different** branches forking from `master`;
-  - PRs with dependencies should be raised only after all prerequisite PRs are merged into `master`.
+  - Consider rebasing your commits to keep the commit history clean and meaningful.
 
-### **PR naming conventions**
+- **If you are making multiple PRs:**
```
Comment (ailzhang) [inline docs/lang/articles/contribution/contributor_guide.md:197]: IMHO this section is not necessary. I hope if we mention it in the future we provide a "better" solution than creating multiple prs from different branches (e.g. ghstack :P https://github.com/ezyang/ghstack)

File: docs/lang/articles/contribution/contributor_guide.md [hunk 13]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
+`[gui] [Mac] Support modifier keys` (only one tag has its initial capitalized)
+:::
 
-- Continuous Integration (CI) will **build** and **test** your commits in a specific PR in multiple environments.
-- CI is triggered each time you push a commit to an open PR.
-- A ✔️ on the left-hand side of a commit hash means that CI passes, whilst a ❌ means that CI fails.
+- **Title naming conventions:**
+  - Keep your PR title short enough but ensure that it carries necessary information.
+  - Do not include back quotes ("\`") in your PR title.
+  - Capitalize the initial letter of your title, which is the word immediately after your tag(s).
```
Comment (ailzhang) [inline docs/lang/articles/contribution/contributor_guide.md:249]: the first letter?

**Review by ailzhang — APPROVED** [review]
LGTM!

File: docs/lang/articles/contribution/contributor_guide.md [hunk 14]
```diff
@@ -22,14 +22,10 @@ Always use straightforward (sometimes even brute-force) solutions: Complicated c
 
 ### Juxtapose pros and cons
 
-When it comes to making a design decision, weigh up its pros and cons. A design is *good to go,* so long as its advantages outweigh its disadvantages.
+When it comes to making a design decision, weigh up its pros and cons. A design is *good to go* so long as its advantages outweigh its disadvantages.
 
 > "Perfection (in design) is achieved not when there is nothing more to add, but rather when there is nothing more to take away." [Antoine de Saint-Exupéry](https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar)
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:27]: While the sentence stands on itself, I feel like its meaning it more like “keep things simple”. The context, however, is about “everything has pros and cons and perfectionism is often not optimal”. WDYT?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 15]
```diff
@@ -43,8 +39,8 @@ Our ultimate goal is to build a sustainable, prosperous Taichi community, and ef
   - Use a spell checker, such as [Grammarly](https://app.grammarly.com/), to improve your writing in terms of grammar, style, and tone.
 
 - Constructive and courteous: Base your feedback and discussions on facts, *NOT* on personal feelings.
-  - Acceptable: *"This design could be confusing to new Taichi users. If it were designed this way, it could..."*
-  - Undesirable: *"This design is terrible."*
+  - Acceptable😃: *"This design could be confusing to new Taichi users. If it were designed this way, it could..."*
+  - Undesirable😞: ~~*"This design is terrible."*~~
 
 ## What can you contribute
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:45]:
```suggestion
## What you can contribute
```
(Not entirely sure about this, so feel free to reject this proposal.)

File: docs/lang/articles/contribution/contributor_guide.md [hunk 16]
```diff
@@ -69,44 +65,43 @@ Except for minor updates, most PRs start from a developer taking over an issue.
 
 | Issue Tag                                                    | Description               | Target developer                               |
 | ------------------------------------------------------------ | ------------------------- | ---------------------------------------------- |
-| [good first issue](https://github.com/taichi-dev/taichi/issues?q=is:open+is:issue+label:"good+first+issue") |                           | Developers new to Taichi                       |
-| [welcome contribution](https://github.com/taichi-dev/taichi/issues?q=is:open+is:issue+label:"welcome+contribution") | Slightly more challenging | Developers who wish to dive deeper into Taichi |
+| [good first issue](https://github.com/taichi-dev/taichi/issues?q=is:open+is:issue+label:"good+first+issue") | Issues that are easy to start with                          | Developers new to Taichi                       |
+| [welcome contribution](https://github.com/taichi-dev/taichi/issues?q=is:open+is:issue+label:"welcome+contribution") | *Slightly* more challenging | Developers who wish to dive deeper into Taichi |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:69]:
```suggestion
| [welcome contribution](https://github.com/taichi-dev/taichi/issues?q=is:open+is:issue+label:"welcome+contribution") | *Slightly* more challenging issues | Developers who wish to dive deeper into Taichi |
```

Shall we consistently use noun phrases in the same column?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 17]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
-  - Ensure that you @ the corresponding developers in your comments or request reviews from them.
-- All PRs should ideally come with corresponding **tests**. See [Testing](./contributor_guide#testing).
-- Most PRs, except for those for internal compiler implementations, ought to be accompanied with **document updates**. See [Documentation](./contributor_guide#documentation).
-- All PRs must pass **continuous integration tests** before they get merged. See [Using continuous integration](./contributor_guide#using-continuous-integration).
-- All PRs must pass **code format checks**. See [Enforcing code style](./contributor_guide#enforcing-code-style).
-- Read a great article from Google on [how to have your PR merged quickly](https://testing.googleblog.com/2017/06/code-health-too-many-comments-on-your.html)
+  - Click **Convert to draft** on your PR page to convert the PR to draft, indicating that you are still working on it.
+  - Click **Ready for review** when you are all set and up for a review.
+  - See the [Draft](https://github.blog/2019-02-14-introducing-draft-pull-requests/) for more information.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:191]:
```suggestion
  - See [GitHub Draft PR Guide](https://github.blog/2019-02-14-introducing-draft-pull-requests/) for more information.
```

"The Draft" sounds a little weird :-)

File: docs/lang/articles/contribution/contributor_guide.md [hunk 18]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
 
-[[PDF]](https://github.com/yuanming-hu/public_files/blob/master/graphics/taichi/google_review_comments.pdf)
+- **If you make multiple commits to an open PR:**
 
-- PR authors are required to **squash commits on your own**.
-- If you are making multiple PRs,
-  - Independent PRs should be based on **different** branches forking from `master`;
-  - PRs with dependencies should be raised only after all prerequisite PRs are merged into `master`.
+  - Consider rebasing your commits to keep the commit history clean and meaningful.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:195]: I feel like in practice we don't really follow this rule. Since we always squash commits in a PR into a single commit with the final commit message being the PR title in the `master` branch, the intermediate commit messages/history are unimportant. Maybe we can remove this bullet?

File: docs/lang/articles/contribution/contributor_guide.md [hunk 19]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
++--->  Prepend at least one tag to your PR title.
+```
 
-- Other PRs (underlying development or intermediate implementation) should use tags **in lowercase letters**, for example`[metal]`, `[vulkan]`, `[ir]`, `[lang]`.
+- **Tag naming conventions:**
+  - Prepend at least one tag, such as `[lang]`, to your PR title.
+  - If you have multiple tags, separate adjacent tags with one space.
+  - See [misc/prtags.json](https://github.com/taichi-dev/taichi/blob/master/misc/prtags.json) for a full list of available tags.
+  - We differentiate PRs for end-users from PRs for developers by *capitalizing tag initial*.
+    - If a PR deals with a feature visible to the end-users, initialize the most relevant tag and the PR will [make it into the release notes](https://github.com/taichi-dev/taichi/blob/master/misc/make_changelog.py). For example, `[Metal]`, `[Vulkan]`, `[IR]`, `[Lang]`, or `[CUDA]` Ensure that your PR title has *AT MOST* one tag dealt this way.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:227]:
```suggestion
    - If a PR deals with a feature visible to the end-users, initialize the most relevant tag and the PR will [make it into the release notes](https://github.com/taichi-dev/taichi/blob/master/misc/make_changelog.py). For example, `[Metal]`, `[Vulkan]`, `[IR]`, `[Lang]`, or `[CUDA]`. Ensure that your PR title has *AT MOST* one tag dealt this way. 
```

File: docs/lang/articles/contribution/contributor_guide.md [hunk 20]
```diff
@@ -131,144 +125,167 @@ python misc/code_format.py
...
 
+When introducing a new tag, ensure that you add it to [misc/prtags.json](https://github.com/taichi-dev/taichi/blob/master/misc/prtags.json) so that others can follow.
 
+:::
 
-## Frequently asked questions
+### PR review & merging checklist
 
-#### What if I didn't format my code style locally?
+Folow this checklist during PR review or merging:
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/contributor_guide.md:276]:
```suggestion
Follow this checklist during PR review or merging:
```

**Review by yuanming-hu — APPROVED** [review]
Fantastic! This chapter looks much improved.

I left some very minor comments. Some of them may make sense, some may not, and whether adopting them is up to you as our technical writing expert :-)

Thank you so much!

File: docs/lang/articles/contribution/contributor_guide.md (hunk 13 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:249]: Initial or initial letter

File: docs/lang/articles/contribution/contributor_guide.md (hunk 14 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:27]: Spot-on! This quote is misplaced. It ought to be placed under ### Be pragmatic & no overkills.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 15 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:45]: I had the same question when I was writing this title. Let's use assertive sentence in a question title.

File: docs/lang/articles/contribution/contributor_guide.md (hunk 16 above)
Comment (Vissidarte-Herman) [reply docs/lang/articles/contribution/contributor_guide.md:69]: Nope, it is not mandatory.
